immuni_analytics/celery/authorization/tasks/authorize_analytics_token.py

```diff
@@ -0,0 +1,163 @@
+
+
+async def _authorize_analytics_token(analytics_token: str, device_token: str) -> None:
+    try:
+        # NOTE: bandit complains about using random.uniform stating "[B311:blacklist] Standard
+        #  pseudo-random generators are not suitable for security/cryptographic purposes.".
+        #  The "waiting for a random time" is not a security/cryptographic action, thus the issue is
+        #  intentionally waived.
+        await _first_step(device_token)
+        await asyncio.sleep(random.uniform(1, config.CHECK_TIME))  # nosec
+        await _second_step(device_token)
+        await asyncio.sleep(random.uniform(1, config.READ_TIME))  # nosec
+        await _third_step(device_token)
+    except DiscardAnalyticsTokenException:
+        return
+
+    await _add_analytics_token_to_redis(analytics_token)
+
+
+async def _first_step(device_token: str) -> None:
+    """
+    Fetch the DeviceCheck bits and ensure the configuration is expected.
+    If not, blacklist the device.
+
+    :raises: DiscardAnalyticsTokenException
+    """
+    device_check_data = await fetch_device_check_bits(device_token)
+    if config.ENV == Environment.RELEASE and device_check_data.used_in_current_month:
+        _LOGGER.warning(
+            "Found token already used in current month",
+            extra=dict(
+                env=config.ENV.value,
+                bit0=device_check_data.bit0,
+                bit1=device_check_data.bit1,
+                last_update_time=device_check_data.last_update_time,
+            ),
+        )
+        raise DiscardAnalyticsTokenException()
+
+    if not device_check_data.is_default_configuration_compliant:
+        _LOGGER.warning(
+            "Found token not default configuration compliant in first step.",
+            extra=dict(
+                env=config.ENV.value,
+                bit0=device_check_data.bit0,
+                bit1=device_check_data.bit1,
+                last_update_time=device_check_data.last_update_time,
+            ),
+        )
+        await _blacklist_device(device_token)
+
+
+async def _second_step(device_token: str) -> None:
+    """
+    Fetch the DeviceCheck bits and ensure the configuration is expected.
+    If it is, set the bits to (0,1) and return.
+    If not, blacklist the device.
+
+    :raises: DiscardAnalyticsTokenException
+    """
+    device_check_data = await fetch_device_check_bits(device_token)
+    if not device_check_data.is_default_configuration_compliant:
+        _LOGGER.warning(
+            "Found token not default configuration compliant in second step.",
+            extra=dict(
+                env=config.ENV.value,
+                bit0=device_check_data.bit0,
+                bit1=device_check_data.bit1,
+                last_update_time=device_check_data.last_update_time,
+            ),
+        )
+        await _blacklist_device(device_token)
+
+    await set_device_check_bits(device_token, bit0=True, bit1=False)
```

Comment:
Checking and setting the device check bits in this way opens up the possibility of concurrency issues.

I recommend using a lock here that is retained before checking the bits and released after setting the bits (simply indent this into a with lock(...) )

Reply:
Discussed and discarded the option